File: java-spring-boot/src/main/java/id/austinpay/webhook/WebhookController.java
package id.austinpay.webhook;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.security.MessageDigest;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.Map;

/**
 * AustinPay Webhook Receiver — Spring Boot
 * ============================================================
 * Secret diambil dari application.properties (austinpay.webhook.secret)
 * atau environment variable WEBHOOK_SECRET.
 */
@RestController
public class WebhookController {

    private final ObjectMapper objectMapper = new ObjectMapper();

    @Value("${austinpay.webhook.secret:whsec_ganti_dengan_secret_anda_dari_dashboard}")
    private String webhookSecret;

    @PostMapping(value = "/webhook", consumes = "application/json")
    public ResponseEntity<Map<String, Object>> webhook(
            @RequestBody byte[] rawBody, // WAJIB byte[], bukan String / @RequestBody DTO — supaya
                                          // Spring tidak melakukan konversi charset atau re-serialize
                                          // yang bisa mengubah bytes sebelum kita hash.
            @RequestHeader(value = "X-AustinPay-Signature", required = false) String signature,
            @RequestHeader(value = "X-AustinPay-Event", required = false) String eventHeader
    ) {
        System.out.println("================ WEBHOOK RECEIVED ================");
        System.out.println("Raw Payload: " + new String(rawBody));
        System.out.println("Signature Header: " + signature);
        System.out.println("Event Header: " + eventHeader);

        if (webhookSecret == null || webhookSecret.isBlank()) {
            return error(HttpStatus.INTERNAL_SERVER_ERROR, "Server misconfiguration: WEBHOOK_SECRET is missing.");
        }

        if (signature == null || signature.isBlank()) {
            System.out.println("Rejected request: X-AustinPay-Signature header is missing.");
            return error(HttpStatus.UNAUTHORIZED, "Unauthorized: Missing signature.");
        }

        String computedSignature;
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(webhookSecret.getBytes(), "HmacSHA256"));
            byte[] computedBytes = mac.doFinal(rawBody);
            computedSignature = HexFormat.of().formatHex(computedBytes);
        } catch (Exception e) {
            System.out.println("Error during signature verification: " + e.getMessage());
            return error(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to compute signature.");
        }

        System.out.println("Signature dari Header : " + signature);
        System.out.println("Signature Hasil Compute: " + computedSignature);

        boolean isValid;
        try {
            byte[] sigBytes = HexFormat.of().parseHex(signature);
            byte[] computedBytes = HexFormat.of().parseHex(computedSignature);
            // MessageDigest.isEqual = constant-time compare sejak JDK 6u17
            isValid = MessageDigest.isEqual(sigBytes, computedBytes);
        } catch (IllegalArgumentException e) {
            isValid = false; // signature header bukan hex string yang valid
        }

        if (!isValid) {
            System.out.println("Rejected request: Invalid signature.");
            return error(HttpStatus.UNAUTHORIZED, "Unauthorized: Invalid signature.");
        }

        JsonNode decoded;
        try {
            decoded = objectMapper.readTree(rawBody);
        } catch (Exception e) {
            return error(HttpStatus.BAD_REQUEST, "Invalid JSON payload.");
        }

        String eventName = eventHeader != null && !eventHeader.isBlank()
                ? eventHeader
                : decoded.path("event").asText("unknown");

        JsonNode payloadData = decoded.has("data") ? decoded.get("data") : decoded;

        System.out.println("Webhook Signature Verified Successfully!");
        System.out.println("Event: " + eventName);
        System.out.println("Payload: " + payloadData);
        System.out.println("==================================================");

        // ── Taruh logika bisnis kamu di sini ──────────────────
        switch (eventName) {
            case "deposit.paid":
                // creditOrderInYourDb(payloadData.get("transactionId").asText(), ...);
                break;
            case "withdraw.approved":
                // markWithdrawSuccess(payloadData.get("transactionId").asText());
                break;
            case "withdraw.rejected":
                // notifyUser(payloadData.get("transactionId").asText(), payloadData.get("reason").asText());
                break;
            default:
                System.out.println("Event tidak dikenal: " + eventName);
        }

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("message", "Webhook received and processed");
        return ResponseEntity.ok(body);
    }

    private ResponseEntity<Map<String, Object>> error(HttpStatus status, String message) {
        Map<String, Object> body = new HashMap<>();
        body.put("error", message);
        return ResponseEntity.status(status).body(body);
    }
}

File: nestjs/src/webhook.controller.ts
/**
 * AustinPay Webhook Receiver - NestJS
 *
 * Webhook Secret didapatkan dari Dashboard AustinPay -> menu Webhook.
 */

import {
  Controller,
  Post,
  Headers,
  Req,
  HttpCode,
  HttpException,
  Logger,
} from '@nestjs/common';
import { Request } from 'express';
import * as crypto from 'crypto';

@Controller()
export class WebhookController {
  private readonly logger = new Logger(WebhookController.name);

  @Post('/webhook')
  @HttpCode(200)
  handleWebhook(
    @Headers('x-austinpay-signature') signature: string,
    @Headers('x-austinpay-event') eventHeader: string,
    @Req() req: Request,
  ) {
    const webhookSecret = process.env.WEBHOOK_SECRET || '';
    // rawBody diaktifkan lewat { rawBody: true } di NestFactory.create (lihat main.ts)
    const rawBody = (req as any).rawBody ? (req as any).rawBody.toString('utf-8') : '';

    this.logger.log('================ WEBHOOK RECEIVED ================');
    this.logger.log(`Event header: ${eventHeader}`);

    let envelope: any = {};
    try {
      envelope = JSON.parse(rawBody);
    } catch { /* ignore */ }

    if (!webhookSecret) {
      throw new HttpException({ error: 'WEBHOOK_SECRET is missing.' }, 500);
    }

    if (!signature) {
      this.logger.warn('Missing X-AustinPay-Signature header');
      throw new HttpException({ error: 'Unauthorized: Missing signature.' }, 401);
    }

    const computedSignature = crypto
      .createHmac('sha256', webhookSecret)
      .update(rawBody)
      .digest('hex');

    const sigBuf = Buffer.from(signature, 'hex');
    const compBuf = Buffer.from(computedSignature, 'hex');

    let isValid = false;
    if (sigBuf.length === compBuf.length) {
      isValid = crypto.timingSafeEqual(sigBuf, compBuf);
    }

    if (!isValid) {
      this.logger.warn('Invalid signature - rejecting request');
      throw new HttpException({ error: 'Unauthorized: Invalid signature.' }, 401);
    }

    this.logger.log('Webhook Signature Verified Successfully!');
    const { transactionId, amount, status } = envelope.data || {};
    this.logger.log(
      `Event: ${envelope.event || eventHeader} | transactionId=${transactionId} amount=${amount} status=${status}`,
    );

    // TODO: taruh logika bisnis kamu di sini berdasarkan envelope.event
    // (deposit.paid, withdraw.approved, withdraw.rejected)

    return { success: true, message: 'Webhook received and processed' };
  }
}
